src/repr/src/row.rs

```diff
@@ -268,23 +250,54 @@ unsafe fn read_untagged_bytes<'a>(data: &'a [u8], offset: &mut usize) -> &'a [u8
 /// and it was only written with a `String` tag if it was indeed UTF-8.
 unsafe fn read_lengthed_datum<'a>(data: &'a [u8], offset: &mut usize, tag: Tag) -> Datum<'a> {
     let len = match tag {
-        Tag::BytesTiny | Tag::StringTiny => read_copy::<u8>(data, offset) as usize,
-        Tag::BytesShort | Tag::StringShort => read_copy::<u16>(data, offset) as usize,
-        Tag::BytesLong | Tag::StringLong => read_copy::<u32>(data, offset) as usize,
-        Tag::BytesHuge | Tag::StringHuge => read_copy::<usize>(data, offset),
+        Tag::BytesTiny | Tag::StringTiny => usize::from(read_byte(data, offset)),
+        Tag::BytesShort | Tag::StringShort => {
+            usize::from(u16::from_le_bytes(read_byte_array(data, offset)))
+        }
+        Tag::BytesLong | Tag::StringLong => {
+            usize::cast_from(u32::from_le_bytes(read_byte_array(data, offset)))
+        }
+        Tag::BytesHuge | Tag::StringHuge => {
+            usize::cast_from(u64::from_le_bytes(read_byte_array(data, offset)))
+        }
         _ => unreachable!(),
     };
     let bytes = &data[*offset..(*offset + len)];
     *offset += len;
     match tag {
         Tag::BytesTiny | Tag::BytesShort | Tag::BytesLong | Tag::BytesHuge => Datum::Bytes(bytes),
         Tag::StringTiny | Tag::StringShort | Tag::StringLong | Tag::StringHuge => {
-            Datum::String(std::str::from_utf8_unchecked(bytes))
+            Datum::String(str::from_utf8_unchecked(bytes))
         }
         _ => unreachable!(),
     }
 }
 
+fn read_byte(data: &[u8], offset: &mut usize) -> u8 {
+    let byte = data[*offset];
+    *offset += 1;
+    byte
+}
+
+fn read_byte_array<const N: usize>(data: &[u8], offset: &mut usize) -> [u8; N] {
+    let mut raw = [0; N];
+    raw.copy_from_slice(&data[*offset..*offset + N]);
+    *offset += N;
+    raw
+}
+
+fn read_date(data: &[u8], offset: &mut usize) -> NaiveDate {
+    let year = i32::from_le_bytes(read_byte_array(data, offset));
+    let ordinal = u32::from_le_bytes(read_byte_array(data, offset));
+    NaiveDate::from_yo(year, ordinal)
```

Comment:
Why not just transmute here? I understand the philosophy of getting rid of unsafe where feasible, but isn’t places where it actually lets you write more performant code _and_ you know for sure it’s correct a valid time to use unsafe?


Is the philosophy here that we want to be robust and fail gracefully even on corrupt stored data?

Reply:
I suppose doing so would make us dependent on NaiveDate’s ABI never changing.